ts/client/src/client.ts

```diff
@@ -3374,64 +3372,230 @@ export class MangoClient {
+      allowCreatingBorrows ?? false,
+      expiryTimestamp,
+    );
+  }
+
+  public async tcsStopLossOnBorrow(
+    group: Group,
+    account: MangoAccount,
+    sellBank: Bank,
+    buyBank: Bank,
+    thresholdPriceUi: number,
+    thresholdPriceInSellPerBuyToken: boolean,
+    maxBuyUi: number | null,
+    maxSellUi: number | null,
+    pricePremium: number | null,
+    allowCreatingDeposits: boolean | null,
+    allowCreatingBorrows: boolean | null,
+    expiryTimestamp: number | null,
+  ): Promise<TransactionSignature> {
+    if (maxBuyUi == null && account.getTokenBalanceUi(sellBank) > 0) {
+      throw new Error(
+        `Only allowed to set stop loss on borrows! Current balance ${account.getTokenBalanceUi(
+          buyBank,
+        )}`,
+      );
+    }
+
+    return await this.tokenConditionalSwapCreate(
+      group,
+      account,
+      sellBank,
+      buyBank,
+      thresholdPriceUi,
+      thresholdPriceInSellPerBuyToken,
+      maxBuyUi ?? -account.getTokenBalanceUi(buyBank),
+      maxSellUi ?? Number.MAX_SAFE_INTEGER,
+      'StopLossOnBorrow',
+      pricePremium,
+      allowCreatingDeposits ?? false,
+      allowCreatingBorrows ?? false,
+      expiryTimestamp,
+    );
+  }
+
+  public async tokenConditionalSwapCreate(
+    group: Group,
+    account: MangoAccount,
+    sellBank: Bank,
+    buyBank: Bank,
+    thresholdPriceUi: number,
+    thresholdPriceInSellPerBuyToken: boolean,
+    maxBuyUi: number,
+    maxSellUi: number,
+    tcsIntention:
+      | 'TakeProfitOnDeposit'
+      | 'StopLossOnDeposit'
+      | 'TakeProfitOnBorrow'
+      | 'StopLossOnBorrow'
+      | null,
+    pricePremium: number | null,
+    allowCreatingDeposits: boolean,
+    allowCreatingBorrows: boolean,
+    expiryTimestamp: number | null,
+  ): Promise<TransactionSignature> {
+    const maxBuy = toNative(maxBuyUi, buyBank.mintDecimals);
+    const maxSell = toNative(maxSellUi, sellBank.mintDecimals);
+
+    if (!thresholdPriceInSellPerBuyToken) {
+      thresholdPriceUi = 1 / thresholdPriceUi;
+    }
+
+    let lowerLimit, upperLimit;
+    const thresholdPrice = toNativeSellPerBuyTokenPrice(
+      thresholdPriceUi,
+      sellBank,
+      buyBank,
     );
-    pricePremiumFraction =
-      pricePremiumFraction > 0 ? pricePremiumFraction : 0.3;
+    const sellTokenPerBuyTokenPrice = buyBank.price
+      .div(sellBank.price)
+      .toNumber();
+
+    if (
+      tcsIntention == 'TakeProfitOnDeposit' ||
+      tcsIntention == 'StopLossOnBorrow' ||
+      (tcsIntention == null && thresholdPrice > sellTokenPerBuyTokenPrice)
+    ) {
+      lowerLimit = thresholdPrice;
+      upperLimit = Number.MAX_SAFE_INTEGER;
+    } else {
+      lowerLimit = 0;
+      upperLimit = thresholdPrice;
+    }
+
     const expiryTimestampBn =
       expiryTimestamp !== null ? new BN(expiryTimestamp) : U64_MAX_BN;
 
+    pricePremium = group.getPriceImpactByTokenIndex(
+      sellBank.tokenIndex,
+      maxSellUi * sellBank.uiPrice,
+    );
```

Comment:
unconditionally overrides the passed-in pricePremium; also, isn't the price impact dependent on both, the sell _and_ buy bank? It looks like this would go quite wrong if you want to buy BONK for USDC and then USDC's price impact gets used.

Reply:
changed to

```

    if (!pricePremium) {
      pricePremium = Math.max(
        group.getPriceImpactByTokenIndex(
          sellBank.tokenIndex,
          maxSellUi * sellBank.uiPrice,
        ),
        group.getPriceImpactByTokenIndex(
          buyBank.tokenIndex,
          maxSellUi * sellBank.uiPrice,
        ),
      );
    }
```

I am thinking of just hardcoding to 10k$ instead of using `maxSellUi * sellBank.uiPrice`, the liquidator would just swap in chunks, or?

Reply:
The price impact is relative to USDC right? Then I'd expect both impacts to be combined (get an approximation of the the full impact by swapping to USDC and then to the target). 

If they return a value in percent, then I'd expect `((1 + sell_impact/100) * (1 + buy_impact/100) - 1) * 100` to give the combined number.

Also the amounts should not both be `maxSellUi * sellBank.uiPrice` -- and actually maxSellUi might be inf. Need to get the min of maxBuy and maxSell (converted to the same token)

Reply:
![Uploading image.png…]()

Reply:
Please see latest change 

```
if (!pricePremium) {
      const buyTokenPriceImpact = group.getPriceImpactByTokenIndex(
        buyBank.tokenIndex,
        5000,
      );
      const sellTokenPriceImpact = group.getPriceImpactByTokenIndex(
        sellBank.tokenIndex,
        5000,
      );
      pricePremium =
        ((1 + buyTokenPriceImpact / 100) * (1 + sellTokenPriceImpact / 100) -
          1) *
        100;
    }
    const pricePremiumFraction = pricePremium > 0 ? pricePremium / 100 : 0.03;
    ```